samza-test/src/test/java/org/apache/samza/storage/kv/TestLargeMessageSafeKeyValueStores.java

```diff
@@ -0,0 +1,240 @@
+            new MapConfig(), false, storeName,
+            new WriteOptions(), new FlushOptions(), keyValueStoreMetrics);
+        break;
+      }
+      default :
+        throw new IllegalArgumentException("Type of store undefined: " + typeOfStore);
+    }
+
+    MessageCollector collector = envelope -> {
+      int messageLength = ((byte[]) envelope.getMessage()).length;
+      if (messageLength > maxMessageSize) {
+        throw new SamzaException("Logged store message size " + messageLength + " for store " + storeName
+            + " was larger than the maximum allowed message size " + maxMessageSize + ".");
+      }
+    };
+    loggedStore = new LoggedStore<>(kvStore, systemStreamPartition, collector, loggedStoreMetrics);
+
+    switch (storeConfig) {
+      case "serde" : {
+        KeyValueStore<byte[], byte[]> largeMessageSafeStore =
+            new LargeMessageSafeStore(loggedStore, storeName, dropLargeMessage, maxMessageSize);
+        store = new SerializedKeyValueStore<>(largeMessageSafeStore, stringSerde, stringSerde,
+            serializedKeyValueStoreMetrics);
+        break;
+      }
+      case "cache-then-serde" : {
+        KeyValueStore<byte[], byte[]> toBeSerializedStore = loggedStore;
+        if (dropLargeMessage) {
+          toBeSerializedStore = new LargeMessageSafeStore(loggedStore, storeName, dropLargeMessage, maxMessageSize);
+        }
+        KeyValueStore<String, String> serializedStore =
+            new SerializedKeyValueStore<>(toBeSerializedStore, stringSerde, stringSerde, serializedKeyValueStoreMetrics);
+        store = new CachedStore<>(serializedStore, cacheSize, batchSize, cachedStoreMetrics);
+        break;
+      }
+      //For this case, the value of dropLargeMessage doesn't matter since we are testing the case when
+      // large messages are expected and StorageConfig.DISALLOW_LARGE_MESSAGES is true.
+      case "serde-then-cache" : {
+        KeyValueStore<byte[], byte[]> cachedStore =
+            new CachedStore<>(loggedStore, cacheSize, batchSize, cachedStoreMetrics);
+        KeyValueStore<byte[], byte[]> largeMessageSafeStore =
+            new LargeMessageSafeStore(cachedStore, storeName, dropLargeMessage, maxMessageSize);
+        store = new SerializedKeyValueStore<>(largeMessageSafeStore, stringSerde, stringSerde,
+            serializedKeyValueStoreMetrics);
+        break;
+      }
+      default :
+        throw new IllegalArgumentException("Store config undefined: " + storeConfig);
+    }
+    store = new NullSafeKeyValueStore<>(store);
+  }
+
+  @After
+  public void teardown() {
+    try {
+      store.close();
+    } catch (SamzaException e) {
+      loggedStore.close();
+    }
+    if (dir != null && dir.listFiles() != null) {
+      for (File file : dir.listFiles())
+        file.delete();
+      dir.delete();
+    }
+  }
+
+  @Test
+  public void testLargeMessagePutFailureForLoggedStoreWithWrappedStore() {
```

Comment:
This test method (and the test methods below) seem to be a bit misleading. For example, this says "put failure", but if "drop large message" is enabled, then there is no expected failure. Also, it isn't really useful to have the "ForLoggedStoreWithWrappedStore" suffix, since not all executions of this test will do that.
Please try to rename them to clarify what they are testing in all cases. For example, this could just be called "testPutLargeMessage".

Reply:
Renamed these methods.